godot-core/src/registry/signal/connect_builder.rs

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use crate::builtin::{Callable, GString, Variant};
+use crate::classes::object::ConnectFlags;
+use crate::meta;
+use crate::obj::{bounds, Bounds, Gd, GodotClass, WithBaseField};
+use crate::registry::signal::{SignalReceiver, TypedSignal};
+
+/// Type-state builder for customizing signal connections.
+///
+/// Allows a high degree of customization for connecting signals, while maintaining complete type safety.
+///
+/// <div class="warning">
+/// <strong>Warning:</strong>
+/// Exact type parameters are subject to change and not part of the public API. We could annotate <code>#[doc(hidden)]</code>, but it would make
+/// things harder to understand. Thus, try not to name the <code>ConnectBuilder</code> type in your code; most connection setup doesn't need it.
+/// </div>
+// If naming the type becomes a requirement, there may be some options:
+// - Use a type alias in the module or TypedSignal, exposing only public parameters. This would work for constructor, but not all transformations.
+// - Pack multiple types together into "type lists", i.e. custom structs carrying the type state. For a user, this would appear as one type,
+// - which could also be #[doc(hidden)]. However, this may make the trait resolution more complex and worsen error messages, so not done now.
+///
+/// # Builder stages
+///
+/// The builder API has a well-defined flow and is separated in stages. In each stage, you have certain builder methods available that you can
+/// or must call, before advancing to the next stage. Check the instructions.
+///
+/// ## Stage 1 (required)
+/// Choose one:
+/// - [`function`][Self::function]: Connect a global/associated function or a closure.
+/// - [`object_self`][Self::object_self]: If you want to connect a method (in stage 2), running on the same object as the signal.
+/// - [`object`][Self::object]: If you want to connect a method, running on a separate object.
+///
+/// ## Stage 2 (conditional)
+/// Required iff _(if and only if)_ `object_self` or `object` was called in stage 1.
```

Comment:
```suggestion
/// Required if _(if and only if)_ `object_self` or `object` was called in stage 1.
```

Reply:
That's deliberate, hence the comment in parentheses:
https://en.m.wikipedia.org/wiki/iff